### ISSUE TITLE
WorldServer crashing to desktop, maybe Update.bat is not finished?

### ISSUE THREAD

**Issue by Kuiriel** [ZhengPeiRu21/mod-individual-progression#96]
I have run the update file with all related files in the installation folder. One command window opened and closed. Another just stayed open - it seems permanently? Your notes indicate it should be a few minutes and on a 13700k I figured that would be accurate.

There is no indication of any activity in task manager by MySQLD.exe or anything else really, which made me think it had finished.

So I closed it.

The error I then got when launching the server was:

Loading Player Create Level HP/Mana Data...
Class 6 Level 1 does not have health/mana data!

The WorldServer then terminates to desktop, and I cannot proceed further.

I am guessing I had to wait longer, so I have launched the Update.bat file again prior to writing this and will give it another half an hour.

If it turns out I've just been impatient, I will close this ticket if it resolves itself.

Edit: Half an hour has passed since I started it. It's remained stuck on the last line this entire time as shown below. There are no other console windows open.

Edit: Still open over 2 hours later. Same error.

```
2023-03-04T07:59:42.876029Z 0 [Warning] TIMESTAMP with implicit DEFAULT value is deprecated. Please use --explicit_defaults_for_timestamp server option (see documentation for more details).
2023-03-04T07:59:42.876066Z 0 [Warning] 'NO_ZERO_DATE', 'NO_ZERO_IN_DATE' and 'ERROR_FOR_DIVISION_BY_ZERO' sql modes should be used with strict mode. They will be merged with strict mode in a future release.
2023-03-04T07:59:42.877498Z 0 [Note] .\Database\MySQL Server 5.7\bin\mysqld.exe (mysqld 5.7.30-log) starting as process 9964 ...
2023-03-04T07:59:42.881469Z 0 [Note] InnoDB: Mutexes and rw_locks use Windows interlocked functions
2023-03-04T07:59:42.881603Z 0 [Note] InnoDB: Uses event mutexes
2023-03-04T07:59:42.881669Z 0 [Note] InnoDB: _mm_lfence() and _mm_sfence() are used for memory barrier
2023-03-04T07:59:42.881729Z 0 [Note] InnoDB: Compressed tables use zlib 1.2.11
2023-03-04T07:59:42.881793Z 0 [Note] InnoDB: Adjusting innodb_buffer_pool_instances from 8 to 1 since innodb_buffer_pool_size is less than 1024 MiB
2023-03-04T07:59:42.882203Z 0 [Note] InnoDB: Number of pools: 1
2023-03-04T07:59:42.882465Z 0 [Note] InnoDB: Not using CPU crc32 instructions
2023-03-04T07:59:42.886478Z 0 [Note] InnoDB: Initializing buffer pool, total size = 8M, instances = 1, chunk size = 8M
2023-03-04T07:59:42.887873Z 0 [Note] InnoDB: Completed initialization of buffer pool
2023-03-04T07:59:42.916419Z 0 [Note] InnoDB: Highest supported file format is Barracuda.
2023-03-04T07:59:42.985782Z 0 [Note] InnoDB: Creating shared tablespace for temporary tables
2023-03-04T07:59:42.986142Z 0 [Note] InnoDB: Setting file '.\ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
2023-03-04T07:59:42.990282Z 0 [Note] InnoDB: File '.\ibtmp1' size is now 12 MB.
2023-03-04T07:59:42.996322Z 0 [Note] InnoDB: 96 redo rollback segment(s) found. 96 redo rollback segment(s) are active.
2023-03-04T07:59:42.996439Z 0 [Note] InnoDB: 32 non-redo rollback segment(s) are active.
2023-03-04T07:59:42.996813Z 0 [Note] InnoDB: Waiting for purge to start
2023-03-04T07:59:43.052632Z 0 [Note] InnoDB: 5.7.30 started; log sequence number 11349332686
2023-03-04T07:59:43.053704Z 0 [Note] InnoDB: Loading buffer pool(s) from C:\Games\WoW\mod-individual-progression\Database\Data\ib_buffer_pool
2023-03-04T07:59:43.054110Z 0 [Note] Plugin 'FEDERATED' is disabled.
2023-03-04T07:59:43.061351Z 0 [Note] Found ca.pem, server-cert.pem and server-key.pem in data directory. Trying to enable SSL support using them.
2023-03-04T07:59:43.061489Z 0 [Note] Skipping generation of SSL certificates as certificate files are present in data directory.
2023-03-04T07:59:43.062243Z 0 [Warning] CA certificate ca.pem is self signed.
2023-03-04T07:59:43.062423Z 0 [Note] Skipping generation of RSA key pair as key files are present in data directory.
2023-03-04T07:59:43.062915Z 0 [Note] Server hostname (bind-address): '*'; port: 3306
2023-03-04T07:59:43.063498Z 0 [Note] IPv6 is available.
2023-03-04T07:59:43.063602Z 0 [Note]   - '::' resolves to '::';
2023-03-04T07:59:43.063717Z 0 [Note] Server socket created on IP: '::'.
2023-03-04T07:59:43.063836Z 0 [Warning] Setting named_pipe_full_access_group='*everyone*' is insecure. Consider using a Windows group with fewer members.
2023-03-04T07:59:43.064160Z 0 [Note] InnoDB: Buffer pool(s) load completed at 230304 18:59:43
2023-03-04T07:59:43.078122Z 0 [Note] Failed to start slave threads for channel ''
2023-03-04T07:59:43.087408Z 0 [Note] Event Scheduler: Loaded 0 events
2023-03-04T07:59:43.087706Z 0 [Note] .\Database\MySQL Server 5.7\bin\mysqld.exe: ready for connections.
Version: '5.7.30-log'  socket: ''  port: 3306  MySQL Community Server (GPL)

```

**Comment by ZhengPeiRu21** [comment]
I am sorry it is not working for you. I just tested the update on a fresh install of the repack and it is work for me. The bat file should start the MySQL server if it is not running, and then run the update files on the correct databases - you should see the message "Now updating the database - please wait!" while the update is processing. The MySQL server window should stay open and continue to run even after the update is complete. Please note that if you are already running the database server before running update, it will use the existing instance - it will open a new instance for a second, see that it is running, and then close that instance.

All the update file does is run the SQL files against the correct databases - if you are familiar with using any MySQL GUI tool like HeidiSQL, you can also use that to run the update files. They will refresh the world database to the correct state (please make sure you run it only on the world database and not your character database so you don't lose any data.)

**Comment by Kuiriel** [comment]
Thank you. Doing that via HeidiSQL has now allowed me to get the world server up and running normally, without the error "Class 6 Level 1 does not have health/mana data!"

The update_auth worked without error, besides complaining about wanting a where clause. Made sure to run it all on world database. update_world worked okay too. 

However, update_character had an error.

SQL Error (1146): Table 'acore_world.character_homebind' doesn't exist.

```
/* Delimiter changed to ; */
/* Connecting to 127.0.0.1 via MariaDB or MySQL (TCP/IP), username root, using password: No ... */
SELECT CONNECTION_ID();
SHOW VARIABLES;
/* Characterset: utf8mb4 */
SHOW /*!50002 GLOBAL */ STATUS;
SELECT NOW();
/* Connected. Thread-ID: 11 */
/* Reading function definitions from C:\Program Files\HeidiSQL\functions-mysql.ini */
SHOW TABLES FROM `information_schema`;
SHOW DATABASES;
SHOW OPEN TABLES FROM acore_world WHERE `in_use`!=0;
USE `acore_world`;
/* Entering session "IndividualProgression" */
SELECT `DEFAULT_COLLATION_NAME` FROM `information_schema`.`SCHEMATA` WHERE `SCHEMA_NAME`='acore_world';
SHOW TABLE STATUS FROM `acore_world`;
SHOW FUNCTION STATUS WHERE `Db`='acore_world';
SHOW PROCEDURE STATUS WHERE `Db`='acore_world';
SHOW TRIGGERS FROM `acore_world`;
SELECT *, EVENT_SCHEMA AS `Db`, EVENT_NAME AS `Name` FROM information_schema.`EVENTS` WHERE `EVENT_SCHEMA`='acore_world';
/* Loading file "C:\Users\UserName\Documents\backupacoreauthqueryipchange.sql" (117 B) into query tab #1 ... */ (I removed my name)
/* Scaling controls to screen DPI: 100% */
/* File "update_auth20230301.sql" executed, with 5 queries and 1 affected rows */
/* SQL Error (1146): Table 'acore_world.character_homebind' doesn't exist */
/* File "update_characters20221017.sql" executed, with 5 queries and 0 affected rows */
/* 2 file(s) processed, in 00:00:47.7 */
```

I tried running the update_characters SQL file on acore_characters instead and received a SQL Error (1060): Duplicate column name 'posO'.

I am able to join the server and find all characters now though.

**Comment by ZhengPeiRu21** [comment]
> Thank you. Doing that via HeidiSQL has now allowed me to get the world server up and running normally, without the error "Class 6 Level 1 does not have health/mana data!"
> 
> The update_auth worked without error, besides complaining about wanting a where clause. Made sure to run it all on world database. update_world worked okay too.
> 
> However, update_character had an error.
> 
> SQL Error (1146): Table 'acore_world.character_homebind' doesn't exist.
> 
> ```
> /* Delimiter changed to ; */
> /* Connecting to 127.0.0.1 via MariaDB or MySQL (TCP/IP), username root, using password: No ... */
> SELECT CONNECTION_ID();
> SHOW VARIABLES;
> /* Characterset: utf8mb4 */
> SHOW /*!50002 GLOBAL */ STATUS;
> SELECT NOW();
> /* Connected. Thread-ID: 11 */
> /* Reading function definitions from C:\Program Files\HeidiSQL\functions-mysql.ini */
> SHOW TABLES FROM `information_schema`;
> SHOW DATABASES;
> SHOW OPEN TABLES FROM acore_world WHERE `in_use`!=0;
> USE `acore_world`;
> /* Entering session "IndividualProgression" */
> SELECT `DEFAULT_COLLATION_NAME` FROM `information_schema`.`SCHEMATA` WHERE `SCHEMA_NAME`='acore_world';
> SHOW TABLE STATUS FROM `acore_world`;
> SHOW FUNCTION STATUS WHERE `Db`='acore_world';
> SHOW PROCEDURE STATUS WHERE `Db`='acore_world';
> SHOW TRIGGERS FROM `acore_world`;
> SELECT *, EVENT_SCHEMA AS `Db`, EVENT_NAME AS `Name` FROM information_schema.`EVENTS` WHERE `EVENT_SCHEMA`='acore_world';
> /* Loading file "C:\Users\UserName\Documents\backupacoreauthqueryipchange.sql" (117 B) into query tab #1 ... */ (I removed my name)
> /* Scaling controls to screen DPI: 100% */
> /* File "update_auth20230301.sql" executed, with 5 queries and 1 affected rows */
> /* SQL Error (1146): Table 'acore_world.character_homebind' doesn't exist */
> /* File "update_characters20221017.sql" executed, with 5 queries and 0 affected rows */
> /* 2 file(s) processed, in 00:00:47.7 */
> ```
> 
> I tried running the update_characters SQL file on acore_characters instead and received a SQL Error (1060): Duplicate column name 'posO'.
> 
> I am able to join the server and find all characters now though.

It looks like your character database already has the necessary updates, so there is no need to run the SQL. I think your update should be finished and working!

**Comment by Kuiriel** [comment]
Thank you!

Do you have any clue as to why I previously received the error? I'm glad it's fixed now though. 

Class 6 Level 1 does not have health/mana data

**Comment by Kuiriel** [comment]
Okay, I'm a fool. So last time I ran the updates, as described, and then I ran the server, and had the error

"Class 6 Level 1 does not have health/mana data!"

So I followed your advice here, and described above how it progressed, but when I launched the server I had clicked the link to the OLD server. So that runs just fine.

The updated server still has the error: "Class 6 Level 1 does not have health/mana data!"

I've tried the whole process again on a backup of the backup, just using HeidiSQL. Same error persists at the end.

I ran Update_World on Acore_World, Update_Characters on Acore_Characters, Update_Auth on Acore_Auth

Update_Characters gave me that SQL Error (1060): Duplicate column name 'posO' - I have no clue of anything else going awry while doing this.

I wonder if this is because I have previously ran an update you had listed here somewhere for correct stat values.

I expect it's simpler for me to save all characters and upload them to a fully updated server.

**Comment by bkmangx1** [comment]
Did you make any changes to DK starting level? Because the message you're seeing is complaining about death knights not having level 1 stats data.
> I expect it's simpler for me to save all characters and upload them to a fully updated server.
Depending on how many characters you have, it might be worth looking into `pdump` command.

**Comment by bkmangx1** [comment]
That error message is complaining about death knight not having stats for level 1, since the class starts at 55 by default. Do you have max character level below 55 or have changed DK starting level?

>  I expect it's simpler for me to save all characters and upload them to a fully updated server.

Depending on how many characters you have, it might be worth looking into pdump command.

**Comment by bkmangx1** [comment]
That error message is complaining about death knight not having stats for level 1, since the class starts at 55 by default. Do you have max character level set below 55 or have changed DK starting level?

**Comment by Kuiriel** [comment]
I did not change those things, but it looks like the update does. 

The World Server does not have this error before patch, and only has data missing and error after update. It looks like level 1 to 54 DK stats were deleted in the update somehow.

For example, in the unpatched database I can see in acore_world a player_classlevelstats table with class, level, basehp and basemana. And it looks like Class 6 level 1 does have health/mana data here - at level 1 it has HP of 22 and mana of 0, and these values increase all the way up to level 80.

In the patched database, level 1 to 54 are missing for class 6.

It is probably faster to copy the old values from pre-patch database and re-insert them via SQL query than it is to backup and re-upload all characters.

Via Google I can see there is some discussion on this and a possible solution here, though I remain unclear on how this has occurred, and I don't have a player_class_stats table that I can see. The below query also adds in other data which I don't presently see fields for in the plater_classlevelstats table.

https://gist.github.com/heyitsbench/21fa68994f0f20563148c164ee3dcc67

Not that it's relevant but just in case, the config files indicate:

StartPlayerLevel = 1
StartHeroicPlayerLevel = 55

In the mod config files it looks normal too
IndividualProgression.DeathKnightStartingProgression = 11

**Comment by ZhengPeiRu21** [comment]
The module does not change stats for Death Knights, and the config does not alter their starting level. I would recommend checking your database for any Death Knight characters and ensure they are all at level 55 or above - it seems the server is trying to load data for a Death Knight below level 55, which is not supported by AzerothCore without custom values being added as of a recent change (https://github.com/azerothcore/azerothcore-wotlk/pull/14378). There has been some discussion about if there is a better way to handle this.

**Comment by Kuiriel** [comment]
Just us five in the family on the server. I'll check again but I am pretty sure we have no death knights and no way to have set their level before 55. All I know is that the database stats were there before the SQL update and gone after.

**Comment by Kuiriel** [comment]
No Death Knights, no class 6 characters. So why is the server trying to load data for a DK below level 55 after the patch?

Something about the recent change to AzerothCore which is then pushed by this update has created a peculiar conflict, but I can't think why on earth my server database is relevant to it.

![noDeathKnights](https://user-images.githubusercontent.com/24553989/224541182-4874b180-cdd7-4d9f-b853-c927ed7e9c16.png)

**Comment by VladonAbaddon** [comment]
I have the same issue and the same error after downloading and installing the update. I don't have any Class 6 characters either. I haven't updated for a while and may have skipped a couple updates

**Comment by ZhengPeiRu21** [comment]
> 

Thank you for the explanation. It sounds like I forgot to include the binaries in the Update archive; every update should always include new binaries. If I ever forgot to include them again, please let me know!

**Comment by Kuiriel** [comment]
Do I need to grab 'each' previous update with SQL contents and apply them sequentially for it to work? Currently I'm presuming the update would work even if I've unwittingly skipped a revision in there.

I tried the most recent update.bat file with all contents of update file in directory, but instead became stuck at 'Added realm "AzerothCore" at 192.168.0.118:8085'

I opened HeidiSQL and ran the queries manually instead.

/* File "update_world20230313.sql" executed, with 1,751 queries and 2,966,923 affected rows */
/* 1 file(s) processed, in 00:00:40.7 */

/* SQL Error (1060): Duplicate column name 'posO' */
/* File "update_characters20230313.sql" executed, with 5 queries and 0 affected rows */
/* 1 file(s) processed, in 00:00:06.9 */

/* File "update_auth20230301.sql" executed, with 5 queries and 1 affected rows */
/* 1 file(s) processed, in 00:00:00.0 */

I'm not sure how much of the database to export via HeidiSQL or how to avoid the errors I saw as it was doing so. I went with: Tools > Export Database as SQL > IndividualProgression ticked > Database Create ticked, Table create ticked, Data: Insert, Filename marked, Export.

**Comment by ZhengPeiRu21** [comment]
It should only be necessary to use the latest update; it should include all previous updates. If this is ever not the case I will mention it in the release notes. This does lead to warnings such as the duplicate column name for posO, but this can be safely ignored. The updates appear to have all run successfully. Do you receive any error messages in the worldserver (not authserver) when starting up?

**Comment by Kuiriel** [comment]
I had to launch worldserver via a separate command window so it wouldn't keep closing while I was trying to read the error, ha.

```
In mysql_stmt_prepare() id: 447, sql: "INSERT IGNORE INTO profanity_name (name) VALUES (?)"
Table 'acore_characters.profanity_name' doesn't exist
Closing down DatabasePool 'acore_characters'.
Asynchronous connections on DatabasePool 'acore_characters' terminated. Proceeding with synchronous connections.
All connections on DatabasePool 'acore_characters' closed.
Could not prepare statements of the Character database, see log for details.
```

Does that help?

**Comment by Kuiriel** [comment]
No worries, that's why we have backups we can play with. I feel bad for bothering you and wasting your time because I have another error!

/* SQL Error (1813): Tablespace '`acore_characters`.`profanity_name`' exists. */
/* Affected rows: 0  Found rows: 0  Warnings: 1  Duration for 1 of 2 queries: 0.000 sec. */

Okay, I see I had to remove profanity_name..ibd that I had added earlier. Restarted. Tried again.

Then I got

/* SQL Error (1050): Table '`acore_characters`.`profanity_name`' already exists */

Then it goes back to the previous error if I run the query again.

I deleted the IBD file again per instructions in worldserver cmd prompt. I searched for it via profanity_name.* in the entire folder and subdir but it's not there. I launch play.bat and HeidiSQL. There is definitely no profanity_name table there right now. I run the query. It says it exists. It still doesn't show up in the list.

2023-03-17T05:54:53.984780Z 0 [Note] InnoDB: Some operating system error numbers are described at http://dev.mysql.com/doc/refman/5.7/en/operating-system-error-codes.html
2023-03-17T05:54:53.984897Z 0 [ERROR] InnoDB: Cannot open datafile for read-only: '.\acore_characters\profanity_name.ibd' OS error: 71
2023-03-17T05:54:53.984969Z 0 [ERROR] InnoDB: Operating system error number 203 in a file operation.
2023-03-17T05:54:53.985040Z 0 [Note] InnoDB: Some operating system error numbers are described at http://dev.mysql.com/doc/refman/5.7/en/operating-system-error-codes.html
2023-03-17T05:54:53.985113Z 0 [ERROR] InnoDB: Could not find a valid tablespace file for `acore_characters/profanity_name`. Please refer to http://dev.mysql.com/doc/refman/5.7/en/innodb-troubleshooting-datadict.html for how to resolve the issue.
2023-03-17T05:54:53.985250Z 0 [Warning] InnoDB: Ignoring tablespace `acore_characters/profanity_name` because it could not be opened.
2023-03-17T05:54:54.005976Z 0 [Note] InnoDB: Removed temporary tablespace data file: "ibtmp1"
2023-03-17T05:54:54.006118Z 0 [Note] InnoDB: Creating shared tablespace for temporary tables
2023-03-17T05:54:54.006402Z 0 [Note] InnoDB: Setting file '.\ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
2023-03-17T05:54:54.009632Z 0 [Note] InnoDB: File '.\ibtmp1' size is now 12 MB.
2023-03-17T05:54:54.015104Z 0 [Note] InnoDB: 96 redo rollback segment(s) found. 96 redo rollback segment(s) are active.
2023-03-17T05:54:54.015262Z 0 [Note] InnoDB: 32 non-redo rollback segment(s) are active.
2023-03-17T05:54:54.015524Z 0 [Note] InnoDB: Waiting for purge to start
2023-03-17T05:54:54.076942Z 0 [Note] InnoDB: 5.7.30 started; log sequence number 11353592584
2023-03-17T05:54:54.077271Z 0 [Note] InnoDB: Loading buffer pool(s) from C:\Games\WoW\mod-individual-progression\Database\Data\ib_buffer_pool
2023-03-17T05:54:54.077275Z 0 [Note] Plugin 'FEDERATED' is disabled.
2023-03-17T05:54:54.078716Z 0 [Note] InnoDB: Buffer pool(s) load completed at 230317 16:54:54
2023-03-17T05:54:54.081412Z 0 [Note] Found ca.pem, server-cert.pem and server-key.pem in data directory. Trying to enable SSL support using them.
2023-03-17T05:54:54.081537Z 0 [Note] Skipping generation of SSL certificates as certificate files are present in data directory.
2023-03-17T05:54:54.082296Z 0 [Warning] CA certificate ca.pem is self signed.
2023-03-17T05:54:54.082495Z 0 [Note] Skipping generation of RSA key pair as key files are present in data directory.
2023-03-17T05:54:54.082718Z 0 [Note] Server hostname (bind-address): '*'; port: 3306
2023-03-17T05:54:54.082902Z 0 [Note] IPv6 is available.
2023-03-17T05:54:54.083001Z 0 [Note]   - '::' resolves to '::';
2023-03-17T05:54:54.083111Z 0 [Note] Server socket created on IP: '::'.
2023-03-17T05:54:54.083228Z 0 [Warning] Setting named_pipe_full_access_group='*everyone*' is insecure. Consider using a Windows group with fewer members.
2023-03-17T05:54:54.094819Z 0 [Note] Failed to start slave threads for channel ''
2023-03-17T05:54:54.101428Z 0 [Note] Event Scheduler: Loaded 0 events
2023-03-17T05:54:54.101665Z 0 [Note] .\Database\MySQL Server 5.7\bin\mysqld.exe: ready for connections.
Version: '5.7.30-log'  socket: ''  port: 3306  MySQL Community Server (GPL)
2023-03-17T05:56:01.488286Z 11 [ERROR] InnoDB: Operating system error number 80 in a file operation.
2023-03-17T05:56:01.488515Z 11 [Note] InnoDB: Some operating system error numbers are described at http://dev.mysql.com/doc/refman/5.7/en/operating-system-error-codes.html
2023-03-17T05:56:01.488698Z 11 [ERROR] InnoDB: Cannot create file '.\acore_characters\profanity_name.ibd'
2023-03-17T05:56:01.488816Z 11 [ERROR] InnoDB: The file '.\acore_characters\profanity_name.ibd' already exists though the corresponding table did not exist in the InnoDB data dictionary. Have you moved InnoDB .ibd files around without using the SQL commands DISCARD TABLESPACE and IMPORT TABLESPACE, or did mysqld crash in the middle of CREATE TABLE? You can resolve the problem by removing the file '.\acore_characters\profanity_name.ibd' under the 'datadir' of MySQL.

**Comment by Kuiriel** [comment]
Success with launching it after trying updating untouched backup, so previously touching IBDs had broken it. However many map errors because generatorv15, expected v16. Fixed by copying over updated files from AzerothCore January release.

I'll mark this as solved, but I saw a separate bug which can go in its own report.

ConditionMgr: Item 49205 not found in LootTemplate 4421
Not handled grouped condition, SourceGroup 4421
